File: src/portadoc/browser/redact.ts
/**
 * Entity detection for browser-side redaction.
 * Mirrors Python implementation in src/portadoc/redact.py
 */

import { getBasePath } from './basePath';

export enum EntityType {
  NONE = '',
  NAME = 'NAME',
  DATE = 'DATE',
  CODE = 'CODE',
}

// DATE patterns
const DATE_PATTERNS = [
  // US date: 7/24/25, 12-31-2024, with optional trailing punctuation
  /^\d{1,2}[/\-]\d{1,2}[/\-]\d{2,4}[,.]?$/,
  // ISO date: 2024-12-31
  /^\d{4}-\d{2}-\d{2}$/,
];

// CODE patterns (numeric identifiers)
const CODE_PATTERNS = [
  // Phone: (555) 123-4567 or 555-123-4567 or 555.123.4567
  /^\(?\d{3}\)?[\s\-.]?\d{3}[\s\-.]?\d{4}$/,
  // SSN: 123-45-6789
  /^\d{3}-\d{2}-\d{4}$/,
  // Generic 5+ digit codes
  /^\d{5,}$/,
  // Alphanumeric codes: ABC12345, A1234567
  /^[A-Z]{1,3}\d{4,}$/,
  // MRN-style: 7829341 (7 digits)
  /^\d{7}$/,
];

// Exclusion patterns (false positives to avoid)
const EXCLUSION_PATTERNS = [
  // Time: 10:28, 2:30
  /^\d{1,2}:\d{2}$/,
  // Currency: $100, $1,000.00
  /^\$[\d,]+\.?\d*$/,
  // Percent: 50%, 3.5%
  /^\d+\.?\d*%$/,
  // Common abbreviations that look like codes
  /^(AM|PM|ID|MRN|DOB|SSN|EIN)$/i,
];

function matchesDate(text: string): boolean {
  return DATE_PATTERNS.some(p => p.test(text));
}

function matchesCode(text: string): boolean {
  return CODE_PATTERNS.some(p => p.test(text));
}

function matchesExclusion(text: string): boolean {
  return EXCLUSION_PATTERNS.some(p => p.test(text));
}

function stripPunctuation(text: string): string {
  return text.replace(/^[^\w]+|[^\w]+$/g, '');
}

export interface EntityDetectorOptions {
  detectNames?: boolean;
  detectDates?: boolean;
  detectCodes?: boolean;
}

export interface DetectionResult {
  entityType: EntityType;
  redact: boolean;
}

/**
 * Fast entity detector using regex and dictionary lookup.
 *
 * Detection order (most specific first):
 * 1. Check exclusions (time, currency, percent) -> skip
 * 2. Check date patterns -> DATE
 * 3. Check code patterns -> CODE
 * 4. Check names dictionary -> NAME
 */
export class EntityDetector {
  private names: Set<string> = new Set();
  private detectNames: boolean;
  private detectDates: boolean;
  private detectCodes: boolean;
  private loaded = false;

  constructor(options: EntityDetectorOptions = {}) {
    this.detectNames = options.detectNames ?? true;
    this.detectDates = options.detectDates ?? true;
    this.detectCodes = options.detectCodes ?? true;
  }

  /**
   * Load the names dictionary from the server.
   * Must be called before detect() if detectNames is true.
   */
  async loadNames(): Promise<void> {
    if (!this.detectNames) {
      this.loaded = true;
      return;
    }

    try {
      const basePath = getBasePath();
      const response = await fetch(`${basePath}data/dictionaries/us_names.txt`);
      if (!response.ok) {
        console.warn('Failed to load names dictionary, name detection disabled');
        this.detectNames = false;
        this.loaded = true;
        return;
      }

      const text = await response.text();
      const lines = text.split('\n');
      for (const line of lines) {
        const name = line.trim();
        if (name && !name.startsWith('#')) {
          this.names.add(name.toLowerCase());
        }
      }

      console.log(`Loaded ${this.names.size} names for entity detection`);
      this.loaded = true;
    } catch (err) {
      console.warn('Error loading names dictionary:', err);
      this.detectNames = false;
      this.loaded = true;
    }
  }

  /**
   * Detect entity type for a single word.
   */
  detect(text: string): DetectionResult {
    if (!text || !text.trim()) {
      return { entityType: EntityType.NONE, redact: false };
    }

    const trimmed = text.trim();

    // Check exclusions first (false positives)
    if (matchesExclusion(trimmed)) {
      return { entityType: EntityType.NONE, redact: false };
    }

    // Check dates
    if (this.detectDates && matchesDate(trimmed)) {
      return { entityType: EntityType.DATE, redact: true };
    }

    // Check codes
    if (this.detectCodes && matchesCode(trimmed)) {
      return { entityType: EntityType.CODE, redact: true };
    }

    // Check names (case-insensitive, strip punctuation)
    if (this.detectNames) {
      const cleanText = stripPunctuation(trimmed);
      if (cleanText && this.names.has(cleanText.toLowerCase())) {
        return { entityType: EntityType.NAME, redact: true };
      }
    }

    return { entityType: EntityType.NONE, redact: false };
  }

  /**
   * Detect entities for a batch of texts.
   */
  detectBatch(texts: string[]): DetectionResult[] {
    return texts.map(text => this.detect(text));
  }

  /**
   * Check if the detector is ready (names loaded).
   */
  isLoaded(): boolean {
    return this.loaded;
  }

  /**
   * Get the count of loaded names.
   */
  getNameCount(): number {
    return this.names.size;
  }
}

File: src/portadoc/redact.py
"""Entity detection and redaction marking for extracted words."""

import csv
import re
from enum import Enum
from pathlib import Path
from typing import TextIO

from .patterns import matches_date, matches_code, matches_exclusion


class EntityType(Enum):
    """Type of detected entity."""
    NONE = ""
    NAME = "NAME"
    DATE = "DATE"
    CODE = "CODE"


# Default path to names dictionary
DEFAULT_NAMES_PATH = Path(__file__).parent.parent.parent / "data" / "dictionaries" / "us_names.txt"


def load_names(path: Path | None = None) -> set[str]:
    """
    Load names from dictionary file into a set for O(1) lookup.

    Args:
        path: Path to names file (one name per line).
              Defaults to data/dictionaries/us_names.txt.

    Returns:
        Set of lowercase names.
    """
    if path is None:
        path = DEFAULT_NAMES_PATH

    if not path.exists():
        return set()

    names = set()
    with open(path, encoding="utf-8") as f:
        for line in f:
            name = line.strip()
            if name and not name.startswith("#"):
                names.add(name.lower())
    return names


def strip_punctuation(text: str) -> str:
    """Remove leading/trailing punctuation for matching."""
    return re.sub(r"^[^\w]+|[^\w]+$", "", text)


class EntityDetector:
    """
    Fast entity detector using regex and dictionary lookup.

    Detection order (most specific first):
    1. Check exclusions (time, currency, percent) -> skip
    2. Check date patterns -> DATE
    3. Check code patterns -> CODE
    4. Check names dictionary -> NAME
    """

    def __init__(
        self,
        names_path: Path | None = None,
        detect_names: bool = True,
        detect_dates: bool = True,
        detect_codes: bool = True,
    ):
        """
        Initialize the entity detector.

        Args:
            names_path: Path to names dictionary file.
            detect_names: Whether to detect name entities.
            detect_dates: Whether to detect date entities.
            detect_codes: Whether to detect code entities.
        """
        self._names: set[str] = set()
        self._detect_names = detect_names
        self._detect_dates = detect_dates
        self._detect_codes = detect_codes

        if detect_names:
            self._names = load_names(names_path)

    def detect(self, text: str) -> tuple[EntityType, bool]:
        """
        Detect entity type for a word.

        Args:
            text: The word text to analyze.

        Returns:
            Tuple of (entity, should_redact).
            - entity: EntityType.NONE if not an entity
            - should_redact: True if the word should be redacted
        """
        if not text or not text.strip():
            return EntityType.NONE, False

        text = text.strip()

        # Check exclusions first (false positives)
        if matches_exclusion(text):
            return EntityType.NONE, False

        # Check dates
        if self._detect_dates and matches_date(text):
            return EntityType.DATE, True

        # Check codes
        if self._detect_codes and matches_code(text):
            return EntityType.CODE, True

        # Check names (case-insensitive, strip punctuation)
        if self._detect_names:
            clean_text = strip_punctuation(text)
            if clean_text and clean_text.lower() in self._names:
                return EntityType.NAME, True

        return EntityType.NONE, False

    def detect_batch(self, texts: list[str]) -> list[tuple[EntityType, bool]]:
        """Detect entities for a batch of texts."""
        return [self.detect(text) for text in texts]


def redact_csv(
    input_csv: Path,
    output_csv: Path | None = None,
    names_path: Path | None = None,
    detect_names: bool = True,
    detect_dates: bool = True,
    detect_codes: bool = True,
) -> int:
    """
    Read a CSV, detect entities, and append entity and redact columns.

    Args:
        input_csv: Path to input CSV (from portadoc extract).
        output_csv: Path to output CSV. If None, overwrites input.
        names_path: Path to custom names dictionary.
        detect_names: Whether to detect name entities.
        detect_dates: Whether to detect date entities.
        detect_codes: Whether to detect code entities.

    Returns:
        Number of words marked for redaction.
    """
    detector = EntityDetector(
        names_path=names_path,
        detect_names=detect_names,
        detect_dates=detect_dates,
        detect_codes=detect_codes,
    )

    # Read input CSV
    rows = []
    fieldnames = []
    with open(input_csv, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        fieldnames = list(reader.fieldnames or [])
        rows = list(reader)

    # Add new columns if not present
    if "entity" not in fieldnames:
        fieldnames.append("entity")
    if "redact" not in fieldnames:
        fieldnames.append("redact")

    # Detect entities and mark redactions
    redact_count = 0
    for row in rows:
        text = row.get("text", "")
        entity, should_redact = detector.detect(text)

        row["entity"] = entity.value
        row["redact"] = "true" if should_redact else "false"

        if should_redact:
            redact_count += 1

    # Write output CSV
    out_path = output_csv if output_csv else input_csv
    with open(out_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)

    return redact_count


def get_redaction_stats(input_csv: Path) -> dict:
    """
    Get statistics about detected entities in a CSV.

    Args:
        input_csv: Path to CSV with entity and redact columns.

    Returns:
        Dictionary with counts by entity type.
    """
    stats = {
        "total_words": 0,
        "redacted_count": 0,
        "by_type": {
            "NAME": 0,
            "DATE": 0,
            "CODE": 0,
        },
    }

    with open(input_csv, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            stats["total_words"] += 1
            if row.get("redact") == "true":
                stats["redacted_count"] += 1
                entity = row.get("entity", "")
                if entity in stats["by_type"]:
                    stats["by_type"][entity] += 1

    return stats
